### PR TITLE
Fix time snap of sliders not matching when SV is not 1.0x

### LINKED ISSUE #21047: Slider placement adjustments don't snap correctly to time values
Likely a regression in the latest release, sliders do not correctly snap to time values when slider velocity is not 1.0x.

### Discussed in https://github.com/ppy/osu/discussions/21046

<div type='discussions-op-text'>

<sup>Originally posted by **Axilotl17** November  1, 2022</sup>
when you try to snap a slider end to a time notch, it bases it off of it was SV 1 not whatever SV it actually is. 
https://user-images.githubusercontent.com/78441708/199162922-5394f3f4-f7e1-43f7-998b-be6ef2407877.mp4

</div>

### PR DESCRIPTION
**Pull request by peppy**
Closes #21047

This regressed with https://github.com/ppy/osu/pull/20850 because the function was used in other places which expect it to factor slider velocity into the equation.

Rather than reverting, I've added a new argument, as based on the method naming alone it was hard to discern whether SV should actually be considered.

The reason for the change in #20850 was to avoid the SV coming in from a reference object which may not have a correct SV in the first place. In such cases, passing `false` to the function will give the expected behaviour.

---

Note that I'm still not decided on whether this should be re-integrated into the distance snap grid, which is why I added this back only the the required paths.

### CODE REVIEW

File: osu.Game/Rulesets/Edit/DistancedHitObjectComposer.cs [hunk 1]
```diff
@@ -273,7 +273,7 @@ public virtual float DurationToDistance(HitObject referenceObject, double durati
         public virtual double DistanceToDuration(HitObject referenceObject, float distance)
         {
             double beatLength = BeatSnapProvider.GetBeatLengthAtTime(referenceObject.StartTime);
-            return distance / GetBeatSnapDistanceAt(referenceObject) * beatLength;
+            return distance / GetBeatSnapDistanceAt(referenceObject) * beatLength * referenceObject.DifficultyControlPoint.SliderVelocity;
```
Comment (bdach) [inline osu.Game/Rulesets/Edit/DistancedHitObjectComposer.cs:276]: This confuses me. `GetBeatSnapDistanceAt()` will now include the reference slider velocity by default here, but this line also adds a multiplication by `referenceObject.DifficultyControlPoint.SliderVelocity` - which as far as I can tell will undo the reference slider velocity inclusion (and be tantamount to calling `GetBeatSnapDistanceAt(..., false)`). However `DurationToDistance()` - that is supposed to be the inverse of this function - does not do any of this. What am I not seeing?

Comment (peppy) [reply osu.Game/Rulesets/Edit/DistancedHitObjectComposer.cs:276]: Yeah something is wrong here. This line should not have been committed (it was my initial approach before adding the new parameter). Unfortunate that the new tests didn't catch this one.

Comment (peppy) [reply osu.Game/Rulesets/Edit/DistancedHitObjectComposer.cs]: Fixed, and added test coverage of the two methods which were missing it...

**Review by bdach — APPROVED** [review]
Appears to be correct, to the best of my understanding

**Pull request merged**